### PR TITLE
Improve Domain#overlaps

### PR DESCRIPTION
**Pull request by raunaqmorarka**
Avoid overhead of computing intersection for checking Domain overlap by short circuiting where possible
Refactor TupleDomainParquetPredicate to use Domain#overlaps

### CODE REVIEW

File: presto-spi/src/main/java/io/prestosql/spi/predicate/EquatableValueSet.java [hunk 1]
```diff
@@ -283,6 +302,16 @@ public String toString(ConnectorSession session)
                 .collect(toLinkedSet());
     }
 
+    private static <T> boolean intersectOverlaps(Set<T> set1, Set<T> set2)
+    {
+        for (T element : set1) {
```
Comment (martint) [inline presto-spi/src/main/java/io/prestosql/spi/predicate/EquatableValueSet.java:307]: Iterate over the set with fewest elements.

File: presto-spi/src/main/java/io/prestosql/spi/predicate/EquatableValueSet.java [hunk 2]
```diff
@@ -243,6 +243,25 @@ else if (otherValueSet.isWhiteList()) {
...
+        if (whiteList && otherValueSet.isWhiteList()) {
+            return intersectOverlaps(entries, otherValueSet.entries);
+        }
+        else if (whiteList) {
+            return !otherValueSet.entries.containsAll(entries);
+        }
+        else if (otherValueSet.isWhiteList()) {
+            return !entries.containsAll(otherValueSet.entries);
+        }
+        else {
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply presto-spi/src/main/java/io/prestosql/spi/predicate/EquatableValueSet.java:260]: I didn't follow this, isn't this else block needed when `whiteList` and `otherValueSet.isWhiteList()` both are false ? `EquatableValueSet#intersect` also has similar if-else conditions.

File: presto-spi/src/main/java/io/prestosql/spi/predicate/EquatableValueSet.java [hunk 3]
```diff
@@ -283,6 +302,16 @@ public String toString(ConnectorSession session)
                 .collect(toLinkedSet());
     }
 
+    private static <T> boolean intersectOverlaps(Set<T> set1, Set<T> set2)
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply presto-spi/src/main/java/io/prestosql/spi/predicate/EquatableValueSet.java]: Done

File: presto-spi/src/main/java/io/prestosql/spi/predicate/EquatableValueSet.java (hunk 1 above)
Comment (raunaqmorarka) [reply presto-spi/src/main/java/io/prestosql/spi/predicate/EquatableValueSet.java:313]: Made this change here and in `intersect` method

**Comment by raunaqmorarka** [comment]
> A couple of minor comments. Also, can you add some tests for the new functionality?

I found that the `overlaps` function is tested extensively in `TestEquatableValueSet`, `TestSortedRangeSet` and `TestDomain`. These are covering all the scenarios I had in mind.

**Pull request merged**